internal/app/machined/pkg/system/runner/process/process.go

```diff
@@ -108,6 +109,12 @@ func (p *processRunner) run(eventSink events.Recorder) error {
 		return errors.Wrap(err, "error building command")
 	}
 
+	notifyCh := make(chan reaper.ProcessInfo, 8)
+	usingReaper := reaper.Notify(notifyCh)
```

Comment:
It looks like `reaper.Notify` returns true only if it's running? I guess I don't understand  when/why we would want it to return `false`? The way I read this now it makes it seem like this is optional and up to the caller to decide if they want to use the reaper. I'm probably not understanding something fully, but this confuses me a little.

Reply:
yep, exactly, it's for the caller to know whether reaper is running or not.

so the thing is that we need to subscribe for notifications before starting a process (as a process might exit before we get control back), so we don't know the PID of the child we're going to create.

if running with reaper, we just wait on the channel and filter for notification, without reaper we do regular `cmd.Wait()` (which does `waitpid(pid, ...)`).

I wanted the code in `pkg/cmd` and process runner to work with and without reaper, as it might be used outside of `machined` (PID 1 process), so I didn't want to have a hard dependency on reaper running, but at the same time flows with reaper and without reaper should be different

Reply:
so if we could wait for notifications once `cmd.Start()` returns, it could have been as easy as `reaper.Wait(cmd.Process.Pid)`, but as we don't know the PID before fork(), this has to be split into subscribe for notifications first, fork(), wait, unsubscribe flow, so code is a bit ugly

Reply:
Ahh ok. Having it work for things outside out machined explains it.